docs/proposals/emptydir-pv.md

```diff
@@ -0,0 +1,151 @@
+
+This document proposes a model for managing emptydir persistent volumes backed by node local storage.
+
+## Abstract
+We currently have:
+* Various different types of "local volumes", like emptyDir, hostPath and gitRepo
+* A way to "claim" storage, such that the pod referencing the claim can continue using the storage across pod/node restarts
+* A way to dynamically provision volumes to match un-bound/pending claims
+
+The second and third features allow us to say: provision a volume on demand, that survives pod/node restarts by using network attached storage (NFS, network block device from cloudprovider, distributed file system like gluster). This proposal is about a new `emptyDir` persistent volume that combines all three features. The idea is that we can sacrifice some durability to get a similar abstraction using *node local storage*. We expect this new PV source to be useful to applications that:
+* Prioritize latency over durability (eg: async replicated database using local ssd)
+* Prioritize cost over durability (eg: dev/test prototyping)
+* Bring their own out of band save/restore snapshotting
+
+## Goals
+* Define a new persistent volume source that is backed by local storage and survives pod restarts
+* Every pod referencing this PV must land on the same node (this concept has been referred to as "data gravity").
+* Once this PV is bound to a node, it cannot get unbound till a human intervenes or the node is deleted
+* The PV must require low user interaction. It should be provisionable dynamically.
+* The directory backing the PV must get deleted when the PV is deleted.
+
+## Non goals
+* A local storage abstraction that survives extended node outages
+* Snapshotting or simply copying the local volume to another node
+
+## A new PV storage source
+
+As described above, the storage source for the PV must be a local volume. While we already allow hostPath as a PersistentVolumeSource, it is insufficient because:
+* user needs to specify the exact path, thus requires knowledge of the filesystem
+* hard to dynamically provision a hostpath
+* hostpath volumes are not cleaned up
+
+The `emptydir` abstraction solves these problems:
+* it is allocated in `/var/lib/kubelet/resource/uid/`, a location that is predictable but unknow till `resource` creation time
+* The node takes care of allocation, a dynamic provisioner just needs to create the PV
+* it is cleaned up when the pod dies (or in this case when the PV is deleted)
+
+We can create a new [PersistetVolumeSource](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L240) for EmptyDir. Users will request storage by creating PersistetVolumeClaims. If the PVC isn't bound to a PV, one will be created dynamically. The user can't control which node the storage for this PV comes from, or where in the node's filesystem it is allocated. Users can also hand create PVs bound to specific nodes (TODO: elaborate on QoS and what happens if there's no space on the node).
```

Comment:
> So just to make sure I understand, you're saying that if the PVC starts out unbound, then the user can't try to "steer" it towards a particular node, but they can create the PV ahead of time bound to a specific node? That's fine (until you get your first dozen requests for a general constraint/preference language for dynamically provisioning local PVs :-) , I just wanted to make sure I understand.

Yeah they *need* to be able to steer a PV to another node to handle the failover scenario. The steps involved there would be: 
1. Pull the drive out of node 1 and mount it as a random path on node2 (or any healthy vm)
2. rsync data from /randome-path/pv-uuid to node2:/kubelet/root-dir/pv-uuid
3. Set the sticky-host annotation to node 2

> The user can't control which node the storage for this PV comes from, or where in the node's filesystem it is allocated. Users can also hand create PVs bound to specific nodes

That line refers to the PV created by the dynamic provisioner. If they turn off the provisioner, they can still create a PV that points to the given node, but they can't choose where on the filesystem the directory is allocated, because it's an emptyDir. 

If they want to choose both the node and the path, they'd use the "stick-host" annotation + a hostPath pv. We already allow hostPath pvs, so all we need to make that use case work is to teach the scheduler to understand "stick-host" on all localVolume types.